src/aof.c

```diff
@@ -1227,24 +1227,36 @@ void flushAppendOnlyFile(int force) {
 
     /* Perform the fsync if needed. */
     if (server.aof_fsync == AOF_FSYNC_ALWAYS) {
-        /* valkey_fsync is defined as fdatasync() for Linux in order to avoid
-         * flushing metadata. */
-        latencyStartMonitor(latency);
-        /* Let's try to get this data on the disk. To guarantee data safe when
-         * the AOF fsync policy is 'always', we should exit if failed to fsync
-         * AOF (see comment next to the exit(1) after write error above). */
-        if (valkey_fsync(server.aof_fd) == -1) {
-            serverLog(LL_WARNING,
-                      "Can't persist AOF for fsync error when the "
-                      "AOF fsync policy is 'always': %s. Exiting...",
-                      strerror(errno));
-            exit(1);
+        /* If user enable io_uring and system support it, give io_uring a chance? */
+        if (server.io_uring_enabled) {
+            int ret = ioUringPrepFsyncAndSubmit(server.io_uring, server.aof_fd);
```

Comment:
if we exhausted the queue depth (256), we would get stalled here too, correct?

Reply:
`#define IO_URING_DEPTH 256` just initialize the submission and completion queues in the kernel with at least **256**, and by default completion queues is twice of submission queue size, the SQ ring entries do not impose a limit on the number of in-flight requests that the ring can support, it merely limits the number that can be submitted to the kernel in one go (batch). It will not get stalled here. Consider we have only 1 request for submission queue, the overflow of completion queue rarely happen.

Reply:
Currently we only do one job at a time, the fsync, and then wait for it, right? So this number is only for future usages?

Reply:
I think so, maybe useful in https://github.com/valkey-io/valkey/pull/112.  
IIRC this is the number used by Jens?

Reply:
got it. I have a question about `ioUringPrepFsyncAndSubmit` now. will leave comment in the function